File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumTeleOp.java
package org.firstinspires.ftc.teamcode;

import com.acmerobotics.dashboard.config.Config;
import com.arcrobotics.ftclib.controller.PController;
import com.arcrobotics.ftclib.controller.PIDFController;
import com.arcrobotics.ftclib.drivebase.MecanumDrive;
import com.arcrobotics.ftclib.hardware.RevIMU;
import com.arcrobotics.ftclib.hardware.motors.Motor;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.Servo;
import com.qualcomm.robotcore.hardware.DcMotor.RunMode;
import com.qualcomm.robotcore.hardware.DcMotorSimple.Direction;
import com.qualcomm.robotcore.hardware.TouchSensor;

@TeleOp
@Config
public final class MecanumTeleOp extends LinearOpMode {
    private Motor frontLeftMotor;
    private Motor backLeftMotor;
    private Motor frontRightMotor;
    private Motor backRightMotor;
    public static boolean FIELD_CENTRIC = false;
    private RevIMU imu;

    private DcMotor liftMotor;
    private Servo dumpServo;

    private DcMotor armExtendMotor;
    private DcMotor armLiftMotor;
    private DcMotor wristMotor;
    private Servo clawServo;

    private TouchSensor liftDownSensor;
    private TouchSensor liftUpSensor;

    private boolean autoDown = false;
    private boolean autoUp = false;

    public void runOpMode() throws InterruptedException {
        frontLeftMotor = new Motor(hardwareMap, "frontLeftMotor");
        frontRightMotor = new Motor(hardwareMap, "frontRightMotor");
        backLeftMotor = new Motor(hardwareMap, "backLeftMotor");
        backRightMotor = new Motor(hardwareMap, "backRightMotor");

        imu = new RevIMU(hardwareMap);
        imu.init();

        liftMotor = hardwareMap.dcMotor.get("liftMotor");
        dumpServo = hardwareMap.servo.get("dumpServo");

        armExtendMotor = hardwareMap.dcMotor.get("armExtendMotor");
        armLiftMotor = hardwareMap.dcMotor.get("armLiftMotor");
        wristMotor = hardwareMap.dcMotor.get("wristMotor");
        clawServo = hardwareMap.servo.get("clawServo");

        liftDownSensor = hardwareMap.touchSensor.get("liftDownSensor");
        liftUpSensor = hardwareMap.touchSensor.get("liftDownSensor");

        backRightMotor.setInverted(true);
        armLiftMotor.setDirection(Direction.REVERSE);
        wristMotor.setDirection(Direction.REVERSE);

        MecanumDrive drive = new MecanumDrive(
                frontLeftMotor,
                frontRightMotor,
                backLeftMotor,
                backRightMotor
        );

        waitForStart();

        if (!this.isStopRequested()) {
            while(this.opModeIsActive()) {
                double y = -this.gamepad1.left_stick_y;
                double x = gamepad1.left_stick_x * 1.1;
                double rx = gamepad1.right_stick_x;

                double denominator = Math.max(Math.abs(y) + Math.abs(x) + Math.abs(rx), 1.0);
                double frontLeftPower = (y + x + rx) / denominator;
                double backLeftPower = (y - x + rx) / denominator;
                double frontRightPower = (y - x - rx) / denominator;
                double backRightPower = (y + x - rx) / denominator;

                double mult = Math.min(1.1 - (double)this.gamepad1.right_trigger, 1.0);

                if (this.gamepad1.left_bumper) {
                    mult = 0.5;
                } else if (this.gamepad1.right_bumper) {
                    mult = 0.25;
                }

                if (gamepad1.a) FIELD_CENTRIC = !FIELD_CENTRIC; sleep(300);
                if (!FIELD_CENTRIC) {
                    drive.driveRobotCentric(x * mult, y * mult, rx * mult);
                } else {
                    drive.driveFieldCentric(x * mult, y * mult, rx * mult, imu.getHeading());
                }

//                frontLeftMotor.setPower(frontLeftPower * mult);
//                backLeftMotor.setPower(backLeftPower * mult);
//                frontRightMotor.setPower(frontRightPower * mult);
//                backRightMotor.setPower(backRightPower * mult);


                armExtendMotor.setPower(gamepad2.left_trigger - gamepad2.right_trigger);
                armLiftMotor.setPower(gamepad2.left_stick_y / 3.0);

                if (gamepad2.a) {
                    dumpServo.setPosition(dumpServo.getPosition() + 0.005);
                } else if (gamepad2.b) {
                    dumpServo.setPosition(dumpServo.getPosition() - 0.005);
                }

                if (gamepad2.x && clawServo.getPosition() < 1.0) {
                    clawServo.setPosition(clawServo.getPosition() + 0.005);
                } else if (gamepad2.y && clawServo.getPosition() > 0.0) {
                    clawServo.setPosition(clawServo.getPosition() - 0.005);
                }

                if (gamepad2.dpad_up) {
                    wristMotor.setPower(1.0);
                } else if (gamepad2.dpad_down) {
                    wristMotor.setPower(-1.0);
                } else {
                    wristMotor.setPower(0.0);
                }

                if (gamepad2.right_bumper) {
                    autoDown = !autoDown;
                    sleep(300);
                }

                if (gamepad2.left_bumper) {
                    autoUp = !autoUp;
                    sleep(300);
                }

                if (autoDown) {
                    liftMotor.setPower(1.0);
                } else if (autoUp) {
                    liftMotor.setPower(-1.0);
                } else {
                    liftMotor.setPower(gamepad2.right_stick_y);
                }

                if (liftDownSensor.isPressed() && autoDown) {
                    liftMotor.setPower(-0.5);
                    sleep(500);
                    liftMotor.setPower(0.0);
                    autoDown = false;
                    autoUp = false;
                }

                if (liftUpSensor.isPressed() && autoUp) {
                    liftMotor.setPower(0.5);
                    sleep(500);
                    liftMotor.setPower(0.0);
                    autoDown = false;
                    autoUp = false;
                }

                if (gamepad1.dpad_left && gamepad2.dpad_left) {
                    requestOpModeStop();
                }

                telemetry.addData("Front Left Power", frontLeftPower);
                telemetry.addData("Back Left Power", backLeftPower);
                telemetry.addData("Front Right Power", frontRightPower);
                telemetry.addData("Back Right Power", backRightPower);
                telemetry.addData("Multiplier", mult);
                telemetry.addData("Bucket Position", dumpServo.getPosition());
                telemetry.addData("Claw Position", clawServo.getPosition());
//                telemetry.addData("Lift Positoin", liftMotor.getCurrentPosition());
                telemetry.addData("Lift Power", liftMotor.getPower());
                telemetry.addData("Lift Down Sensor", liftDownSensor.isPressed());
                telemetry.addData("Lift Up Sensor", liftUpSensor.isPressed());
                telemetry.addData("Field Centric", FIELD_CENTRIC);
                telemetry.update();
            }
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumTeleOp.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.Servo
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

@TeleOp
class MecanumTeleOp : LinearOpMode() {
    @Throws(InterruptedException::class)
    override fun runOpMode() {
        // Declare our motors
        // Make sure your ID's match your configuration
        val frontLeftMotor = hardwareMap.dcMotor["frontLeftMotor"]
        val backLeftMotor = hardwareMap.dcMotor["backLeftMotor"]
        val frontRightMotor = hardwareMap.dcMotor["frontRightMotor"]
        val backRightMotor = hardwareMap.dcMotor["backRightMotor"]

        val liftMotor = hardwareMap.dcMotor["liftMotor"]
        val dumpServo = hardwareMap.servo["dumpServo"]

        val armExtendMotor = hardwareMap.dcMotor["armExtendMotor"]
        val armLiftLeftMotor = hardwareMap.dcMotor["armLiftLeftMotor"]
        val armLiftRightMotor = hardwareMap.dcMotor["armLiftRightMotor"]

        val wristServo = hardwareMap.crservo["wristServo"]
        val clawServo = hardwareMap.servo["clawServo"]

        // Reverse the right side motors. This may be wrong for your setup.
        // If your robot moves backwards when commanded to go forwards,
        // reverse the left side instead.
        // See the note about this earlier on this page.
        backRightMotor.direction = DcMotorSimple.Direction.REVERSE

        armLiftRightMotor.direction = DcMotorSimple.Direction.REVERSE

        waitForStart()

        if (isStopRequested) return

        while (opModeIsActive()) {
            val y = -gamepad1.left_stick_y.toDouble() // Remember, Y stick value is reversed
            val x = gamepad1.left_stick_x * 1.1 // Counteract imperfect strafing
            val rx = gamepad1.right_stick_x.toDouble()

            // Denominator is the largest motor power (absolute value) or 1
            // This ensures all the powers maintain the same ratio,
            // but only if at least one is out of the range [-1, 1]
            val denominator = max(abs(y) + abs(x) + abs(rx), 1.0)
            val frontLeftPower = (y + x + rx) / denominator
            val backLeftPower = (y - x + rx) / denominator
            val frontRightPower = (y - x - rx) / denominator
            val backRightPower = (y + x - rx) / denominator

            var mult = min(1.1 - gamepad1.right_trigger.toDouble(), 1.0)
            if (gamepad1.left_bumper) mult = 0.5
            else if (gamepad1.right_bumper) mult = 0.25

            frontLeftMotor.power = frontLeftPower * mult
            backLeftMotor.power = backLeftPower * mult
            frontRightMotor.power = frontRightPower * mult
            backRightMotor.power = backRightPower * mult

            liftMotor.power = gamepad2.right_stick_y.toDouble() / if (gamepad2.right_stick_y < 0) 1 else 5
            armExtendMotor.power = gamepad2.left_trigger.toDouble() - gamepad2.right_trigger.toDouble()
            armLiftLeftMotor.power = gamepad2.left_stick_y.toDouble()
            armLiftRightMotor.power = gamepad2.left_stick_y.toDouble()

            if (gamepad2.a) {
                dumpServo.position += 0.005
            } else if (gamepad2.b) {
                dumpServo.position -= 0.005
            }

            if (gamepad2.x && clawServo.position < 1.0) {
                clawServo.position += 0.001
            } else if (gamepad2.y && clawServo.position > 0.0) {
                clawServo.position -= 0.001
            }

            if (gamepad2.dpad_up) {
                wristServo.power = 1.0
            } else if (gamepad2.dpad_down) {
                wristServo.power = -1.0
            } else {
                wristServo.power = 0.0
            }

            telemetry.addData("Front Left Power", frontLeftPower)
            telemetry.addData("Back Left Power", backLeftPower)
            telemetry.addData("Front Right Power", frontRightPower)
            telemetry.addData("Back Right Power", backRightPower)
            telemetry.addData("Multiplier", mult)
            telemetry.addData("Position", dumpServo.position)
            telemetry.update()
        }
    }
}
